### DISCUSSION TITLE
Improvements to use this as backend plugin.

### DISCUSSION THREAD

**Discussion by robinbonnes** [OFFLINE-GmbH/oc-boxes-support#61] [Ideas and feature requests]
First of all thanks for the great plugin. We are using your plugin in our own plugin to user-friendly generate a quotation or newsletter within the OctoberCMS backend. I really like the way we can implement the boxes editor in forms, split the boxes per different form field with contexts and prefill the editor with several boxes with the page template functionality.

For us, we don't use the plugin for the frontend at all. Maybe we have some improvements which can possibly be implemented in the future:

- Make a setting (or by default) to keep the structure pane open on the left.
- Partial context not working by default for some reason. Had to add: ```window.document.addEventListener('boxes.box.add', function(e) { e.detail['partial_contexts'] = ['name_of_context']; });``` (However then it still doesn't work from within the structure pane)
- Third-party-boxes not working in combination with page templates (plugin can't find the boxes).
- For now we have a generic theme to make the plugin work. In the future we might want to use themes for a frontend website. Maybe there is a way to specify a theme in the field config or make the editor work without using themes at all. For example override boxes.yaml from within a plugin. Use custom HTML for the layout, and don't use themes/multisite.

**Comment by tobias-kuendig** [comment]
Thank you for this feedback @robinbonnes! Using the plugin only as a backend editor is actually a use-case I did not think about yet! But your use-case sounds completely valid. 

The plugin currently has a few dependencies on some of the frontend building blocks, like the layouts and the cms router. I will investiage what it would mean to remove these dependencies. But my first instinct is, that this will complicate the plugin and the implementation on your side: Right now we can use the request handling of October's CMS module to generate a page, add components and assets and render the boxes. All this happens while still considering all of October's events and hooks to modify these processes. Without a proper theme, we will lose all of this.

> Make a setting (or by default) to keep the structure pane open on the left.

This is definitely something we can implement. I'll add this to our backlog.

> Partial context not working by default for some reason. 

Can you try again after updating to version `2.2.11`? We've had a fix for this pending in an unreleased branch. I went ahead and released it.

> Third-party-boxes not working in combination with page templates (plugin can't find the boxes).

I cannot reproduce this issue. I can use 3rd-party partials in a page template just fine.

Looking at the code, this feature requires an active frontend theme, otherwise the registration event is never called. Can you confirm that the problem still exists with an active theme? If so, can you give me any more details about your setup?

**Comment by robinbonnes** [comment]
Thanks for your fast response. 

I agree with losing functionality when we remove the theme part. However it could make sense to wrap the boxes content just in some raw HTML. For some of our use-cases, such as building a newsletter we just need a simple "Text", "Button", "Image" and "Linebreak" box. For this it would be great if we could use the boxes as this:

```
newsletter_content:
    span: adaptive
    type: boxes
    previewHtml: [path/to/surrounding/html]
    template: newsletter
    partialContexts:
        - newsletter
```

The html layout would look like:
```
<!DOCTYPE html>
<html lang="nl">
    <head>
        <title>{{ newsletter_title }}</title>
    </head>

    <body>
        {{ header_content|raw }}

        <div class="wrapper">
            {% page %}
        </div>

        {{ footer_content|raw }}
    </body>
</html>
```

> This is definitely something we can implement. I'll add this to our backlog.

That would be great! 😁

> Can you try again after updating to version `2.2.11`? We've had a fix for this pending in an unreleased branch. I went ahead and released it.

Thanks, I can indeed confirm it is working now. 😁

> Looking at the code, this feature requires an active frontend theme, otherwise the registration event is never called. Can you confirm that the problem still exists with an active theme? If so, can you give me any more details about your setup?

I do have an active theme configured. Also it does recognize the boxes in the editor. However when I fill the themes' `boxes.yaml` with some boxes that are located in the plugin, it does give this error message:
```
An exception has been thrown during the rendering of a template ("cms::lang.cms_object.invalid_file") in "C:\laragon\www\octobercms\plugins/offline/boxes/components/boxespageeditor/render.htm" at line 1.
File: C:\laragon\www\octobercms\vendor\twig\twig\src\Template.php
Line: 408
```

But this doesn't have that much priority, because we can keep the boxes in the theme.
